File: iosgl/iosgl/bt/display/display3d/Display3DSprite.m
//
//  Display3DSprite.m
//  iosgl
//
//  Created by zhao on 7/12/2019.
//  Copyright © 2019 zhao. All rights reserved.
//

#import "Display3DSprite.h"
#import "MaterialManager.h"
#import "ObjDataManager.h"
#import "MetalMatrixUtilities.h"
#import "Scene3D.h"
#import "TimeUtil.h"
#import "Material.h"
#import "Scene_data.h"
#import "MaterialShader.h"
#import "GL_Header.h"
#import "DynamicTexItem.h"
#import "Display3DShader.h"
#import "ProgrmaManager.h"
#import "ObjDataManager.h"


@interface Display3DSprite()
@property (nonatomic, strong) Matrix3D *bindMatrix;
@property (nonatomic, strong) id<IBind> bindTarget;
@property (nonatomic, strong) NSString *bindSocket;
@property (nonatomic, strong) Matrix3D *groupMatrix;
@property (nonatomic, strong) Matrix3D *groupRotationMatrix;
@property(nonatomic,strong)NSString* materialUrl;
@property(nonatomic,strong)Material* material;
@property(nonatomic,strong)MaterialBaseParam* materialParam;
@property (nonatomic, assign) float time;
@property (nonatomic, assign) BOOL isInGroup;
 


@end
@implementation Display3DSprite

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self initData];
    }
    return self;
}
-(void)initData;
{
    [self onCreated];

    
        self.textureRes=[[MaterialManager default] getMaterialByUrl:@"tu001.jpg"];
}

-(void)onCreated;
{
  
}
-(void)setMaterialUrl:(NSString*)value  paramData:(NSArray*)paramData;
{
    Display3DSprite* this=self;
    value= [value stringByReplacingOccurrencesOfString:@"_byte.txt" withString:@".txt"];
    value= [value stringByReplacingOccurrencesOfString:@".txt" withString:@"_byte.txt"];
    this.materialUrl =   value;
    
    [[MaterialManager default]getMaterialByte:[[Scene_data default]getWorkUrlByFilePath:value ] fun:^(NSObject *obj) {
        this.material=(Material*)obj;
        if (this.material.useNormal) {
        }
        if (paramData) {
            this.materialParam = [[MaterialBaseParam alloc]init];
            [this.materialParam setData:this.material ary:paramData];
         }
    } info:nil autoReg:YES regName:MaterialShader.shaderStr shader3DCls:[[MaterialShader alloc]init]];
}
-(void)setGroup:(Vector3D*)pos rotaion:(Vector3D*)rotaion  scale:(Vector3D*)scale;
{
    
}
 
-(void)setObjUrl:(NSString*)value;
{
    [[ObjDataManager default]getObjData:value fun:^(ObjData * obj) {
     self.objData=obj;
        [self.objData upToGpu];
 
    }];
}
-(void)setBind:(id<IBind>)bindTarget bindSocket:(NSString*)bindSocket;
{
    self.bindTarget = bindTarget;
    self.bindSocket = bindSocket;
    self.bindMatrix = [[Matrix3D alloc]init];
}
-(void)updateBind;
{
    Display3DSprite* this=self;
    if (this.bindTarget) {
        
        [this.posMatrix3d identity];
        [this.posMatrix3d  appendScale:this.scaleX y:this.scaleX z:this.scaleX];
        if (this.isInGroup) {
            [this.posMatrix3d append:self.groupMatrix];
        }
        [this.bindTarget getSocket:this.bindSocket resultMatrix:this.bindMatrix];
        [this.posMatrix3d append:this.bindMatrix];
        [this.posMatrix3d copyTo:this.rotationMatrix3D];
        [this.rotationMatrix3D identityPostion ];
        if (this.isInGroup) {
            [this.rotationMatrix3D prepend:self.groupRotationMatrix];
            
        }
        
    }
}
-(void)loadObjDataByUrl:(NSString*)url
{
    [[ ObjDataManager default] getObjDataByUrl: url Block:^(ObjData *objData) {
      //  self.objData=objData;
      //  [self.objData upToGpu];
    }];
}

-(void)loadTextureResByUrl:(NSString*)value;
{
    self.textureRes=[[MaterialManager default] getMaterialByUrl:value];
}
-(void)upFrame{
 
    if (self.dynamic) {
        if (!self.sceneVisible) {
            return;
        }
    }
    
    if(self.material){
         [self updateMaterial];
    }else{
        if(self.shader3d&&self.objData){
              GLuint progame= self.shader3d.program;
              glUseProgram(progame);
              
              [self updateBind];
              [self setVc];
              [self setVa];
              [self resetVa];
          }
    }
 
   
    
 
     
    
}
-(void)updateMaterial;
{
    Display3DSprite* this=self;
    if (!this.material || !this.objData) {
        return;
    }
    [this updateBind];
    this.shader3d=this.material.shader;
    
    
    GLuint progame= self.shader3d.program;
    glUseProgram(progame);
    
    [this updateBind];
    [this setVc];
    [this setBaseMaterialVc:this.material];
    [this setMaterialTexture:this.material mp:this.materialParam];
    [this setMaterialVc:this.material mp:this.materialParam];
    [this setMaterialVa];
    [this resetVa];
 
 
}
/*
 基础vc[x]数据
 */
-(void)setBaseMaterialVc:(Material*)material;
{
    Display3DSprite* this=self;
    float t = 0;
    if (material.hasTime) {
        // t = ( [[TimeUtil default]getTimer] - this.time) % 100000 * 0.001;
        t=[[TimeUtil default]getTimer] - this.time;
    }
    if (material.hasTime || material.usePbr || material.useKill) {
        //Scene_data.context3D.setVc4fv($material.shader, "fc0", [1, 0, $material.killNum, t]);//sceneEvnScale,null,killNum,time;
    }
    if (material.scaleLightMap) {
        // Scene_data.context3D.setVcFloat($material.shader, "scalelight", Scene_data.scaleLight);
    }
    if (material.usePbr || material.fogMode == 1) {
        // this.setCamPos($material);
    }
    if (material.fogMode != 0) {
        // Scene_data.context3D.setVc2fv($material.shader, "fogdata", Scene_data.fogData);
        //Scene_data.context3D.setVc3fv($material.shader, "fogcolor", Scene_data.fogColor);
    }
}
-(void)setMaterialVa;
{
    Display3DSprite* this=self;
    Context3D *ctx=this.scene3d.context3D;
    [ctx pushVa:this.objData.verticesBuffer];
    [ctx setVaOffset:this.shader3d name:"v3Position" dataWidth:3 stride:0 offset:0];
    [ctx pushVa:this.objData.uvBuffer];
    [ctx setVaOffset:this.shader3d name:"v2CubeTexST" dataWidth:2 stride:0 offset:0];
    if (this.material.usePbr || this.material.directLight) {
        [ctx pushVa:this.objData.nrmsBuffer];
        [ctx setVaOffset:this.shader3d name:"v3Normal" dataWidth:4 stride:0 offset:0];
        
    }
    [ctx drawCall:self.objData.indexBuffer  numTril:self.objData.trinum ];
}
- (void)resetVa;
{
   Context3D *ctx=self.scene3d.context3D;
    [ctx clearVa:0];
    [ctx clearVa:1];
    [ctx clearVa:2];
}
 /*
 材质vc[x]数据
 */
-(void)setMaterialVc:(Material*)material  mp:(MaterialBaseParam*)mp;
{
    Display3DSprite* this=self;
    Context3D *ctx=self.scene3d.context3D;
    if (material.fcNum <= 0) {
        return;
    }
    float t   = 0;
    if (material.hasTime) {
        t=[[TimeUtil default]getTimer]-this.time;
    }
    [material update:t];
    if (mp) {
        [mp update];
    }
    NSMutableArray<NSNumber*>*   fcData= material.fcData;
    GLfloat fcDataGlArr[fcData.count];
    for (int i=0; i<fcData.count; i++) {
        fcDataGlArr[i]=fcData[i].floatValue;
    }
    fcDataGlArr[0]=0.0;
    fcDataGlArr[1]=1.7677669525146484;
    fcDataGlArr[2]=-1.7677669525146484;
    fcDataGlArr[3]=0.0;
    fcDataGlArr[4]=0.0;
    fcDataGlArr[5]=0.0;
    fcDataGlArr[6]=0.0;
    fcDataGlArr[7]=0.0;
    
    [ctx setVc4fv:material.shader name:"fc" data:fcDataGlArr len:material.fcNum];
    
//    0: 0
//    1: 1.7677669525146484
//    2: -1.7677669525146484
//    3: 0
//    4: 0
//    5: 0
//    6: 0
//    7: 0
    /*
     var t: number = 0;
          if ($material.hasTime) {
              t = (TimeUtil.getTimer() - this.time) % 100000 * 0.001;
          }

          $material.update(t);

          this.setCamPos($material);

          if ($mp) {
              $mp.update();
          }
 
     */
}
 
-(void)setMaterialTexture:(Material*)material  mp:(MaterialBaseParam*)mp;
{
    Context3D *ctx=self.scene3d.context3D;
    NSArray<TexItem*>* texVec  = mp.material.texList;
    TexItem* texItem;
    for (int i   = 0; i < texVec.count; i++) {
        texItem=texVec[i];
        if (texItem.isDynamic) {
            continue;
        }
        if (texItem.type == TexItem.LIGHTMAP) {
            // Scene_data.context3D.setRenderTexture($material.shader, texVec[i].name, this.lightMapTexture, texVec[i].id);
            NSLog(@"TexItem.LIGHTMAP)");
        }
        else if (texItem.type == TexItem.LTUMAP && [Scene_data default].pubLut ) {
            //  Scene_data.context3D.setRenderTexture($material.shader, texVec[i].name, Scene_data.pubLut, texVec[i].id);
             NSLog(@"TexItem.LTUMAP)");
        }
        else if (texItem.type == TexItem.CUBEMAP) {
            if (material.useDynamicIBL) {// && _reflectionTextureVo) {
                NSLog(@"TexItem.useDynamicIBL)");
            } else {
                if([Scene_data default].skyCubeTexture){
                     [ctx setRenderTextureCube:material.shader name:texItem.name texture:[Scene_data default].skyCubeTexture level:texItem.id];
                }
            }
        }else{
            [ctx setRenderTexture:material.shader name:texItem.name texture:  texItem.textureRes.textTureLuint level:texItem.id];
            
        }
    }
    NSArray<DynamicTexItem*>* texDynamicVec  =( NSArray<DynamicTexItem*>*) mp.dynamicTexList;
    for (int i   = 0; i < texDynamicVec.count; i++) {
        texItem=texDynamicVec[i].target;
        if(texItem ){
            [ctx setRenderTexture:material.shader name:texItem.name  texture:texDynamicVec[i].textureRes.textTureLuint level:texItem.id];
            
            
        }
    }
    
}

-(void)setVc;
{
    Display3DSprite* this=self;
    Context3D *context3D=this.scene3d.context3D;
    [context3D setVcMatrix4fv:this.shader3d name:"vpMatrix3D" data:this.viewMatrix.m];
    [context3D setVcMatrix4fv:this.shader3d name:"posMatrix3D" data:this.posMatrix3d.m];
    if (this.material.usePbr || this.material.directLight) {
         [context3D setVcMatrix3fv:this.shader3d name:"rotationMatrix3D" data:this.rotationMatrix3D.rotationM];
    }
}
-(void)setVa;
{
     
    
}
 
@end

File: Phnom/andriodGL/app/src/main/java/z3d/display/Display3DSprite.java
package z3d.display;
import android.util.Log;
import java.util.List;
import z3d.base.ObjData;
import z3d.base.ObjDataBackFun;
import z3d.base.ObjDataManager;
import z3d.base.Scene_data;
import z3d.core.Context3D;
import z3d.display.interfaces.IBind;
import z3d.display.role.Display3dMovie;
import z3d.material.DynamicBaseTexItem;
import z3d.material.Material;
import z3d.material.MaterialBackFun;
import z3d.material.MaterialBaseParam;
import z3d.material.MaterialManager;
import z3d.material.TexItem;
import z3d.material.TextureRes;
import z3d.program.MaterialShader;
import z3d.program.Shader3D;
import z3d.res.MaterialInfoVo;
import z3d.scene.Scene3D;
import z3d.units.TimeUtil;
import z3d.vo.Matrix3D;
import z3d.vo.Vector3D;

public   class Display3DSprite extends Display3D {
    private static final String TAG="Display3DSprite";
    public Material material;
    public MaterialBaseParam materialParam;
    public Shader3D shader3D;
    public ObjData objData;
    public Matrix3D modeMatrix;
    public Matrix3D rotationMatrix;
    public boolean dynamic;
    public float time;
    private Vector3D _groupPos;
    private Vector3D _groupRotation;
    private Vector3D _groupScale;

    public Display3DSprite(Scene3D val){
        super(val);
        this.time=0;
        this.modeMatrix=new Matrix3D();
        this.rotationMatrix=new Matrix3D();
    }
    protected TextureRes getMainTextureRes(){
        TexItem texItem  =this.material.getMainTexItem();
        if(texItem==null){
            return  null;
        }
        List<DynamicBaseTexItem> texDynamicVec  =  this.materialParam.dynamicTexList;
        for (int i   = 0; i < texDynamicVec.size(); i++) {
            DynamicBaseTexItem dynamicBaseTexItem=texDynamicVec.get(i);
            if(texItem.paramName.equals(dynamicBaseTexItem.paramName)){
                return dynamicBaseTexItem.textureRes;
            }
        }
        return texItem.textureRes;
    }
    public void upData() {
        super.upData();
        Context3D ctx = this.scene3d.context3D;
        if (this.material != null) {
            this.updateMaterial();
        }
    }
    public void setMaterialUrl(String url, final List<MaterialInfoVo> paramData)
    {
        MaterialManager.getInstance().getMaterialByte(url, new MaterialBackFun() {
            @Override
            public void Bfun(Material value) {
                material=value;
                if(paramData!=null){
                    materialParam=new MaterialBaseParam();
                    materialParam.setData(material,paramData);
                }
            }
        },true, MaterialShader.shaderNameStr,new MaterialShader());
    }
    public void  setObjUrl(String value)
    {
        Log.d(TAG, "value: "+value);
        ObjDataManager.getInstance().getObjData(value, new ObjDataBackFun() {
            @Override
            public void Bfun(ObjData value) {
                objData=value;
            }
        });
    }
    public  void  updateMaterial()
    {
        if(this.material==null||this.objData==null)
        {
            return;
        }
        this.shader3D=this.material.shader;
        Context3D ctx=this.scene3d.context3D;
        ctx.setProgame(this.shader3D.program);
        this.updateBind();
        this.setVc();
        this.setMaterialTexture(this.material,this.materialParam);
        this.setMaterialVc(this.material,this.materialParam);
        this.setMaterialVa();


    }
    public IBind bindTarget;
    public Matrix3D groupMatrix;
    public Matrix3D bindMatrix;
    public Matrix3D groupRotationMatrix;
    private boolean _isInGroup;
    private String bindSocket;
    protected void updateBind()
    {

        if (this.bindTarget!=null) {
            this.posMatrix3d.identity();
            this.posMatrix3d.appendScale(this.scaleX, this.scaleY, this.scaleX);
            if (this._isInGroup) {
                this.posMatrix3d.append(this.groupMatrix);

            }
            this.bindTarget.getSocket(this.bindSocket, this.bindMatrix);
            this.posMatrix3d.append(this.bindMatrix);
            this.bindMatrix.copyTo(this.rotationMatrix);
            this.rotationMatrix.identityPostion();
            if (this._isInGroup) {
                this.rotationMatrix.prepend(this.groupRotationMatrix);
            }

//            this.sceneVisible = (<any>this.bindTarget).visible;
        }

    }
    protected void setVc()
    {
        Context3D ctx=this.scene3d.context3D;
        ctx.setVcMatrix4fv(this.shader3D,"vpMatrix3D",this.scene3d.camera3D.modelMatrix.m);
        ctx.setVcMatrix4fv(this.shader3D,"posMatrix3D",this.posMatrix3d.m);
        if (this.material.usePbr || this.material.directLight) {
            float[] m = new float[9];
            this.rotationMatrix.getRotaion(m);
            ctx.setVcMatrix3fv(this.shader3D,"rotationMatrix3D",m);

        }
    }

    protected void setMaterialVa()
    {
        Context3D ctx=this.scene3d.context3D;
        ctx.setVa(this.shader3D,"v3Position",3,this.objData.vertexBuffer);
        ctx.setVa(this.shader3D,"v2CubeTexST",2,this.objData.uvBuffer);
        if (!(this.material.directLight || this.material.noLight)) {
            ctx.setVa(this.shader3D,"v2lightuv",2,this.objData.lightUvBuffer);
        }
        ctx.drawCall(this.objData.indexBuffer,this.objData.treNum);
    }


    protected void setMaterialVc(Material material,MaterialBaseParam mp)
    {
        if (material.fcNum <= 0) {
            return;
        }
        float t = 0;
        if (material.hasTime) {
            t =  (TimeUtil.getTimer() - this.time)% 100000 * 0.001f;
        }
        material.update(t);
        this.setCamPos(material);
        if (mp!=null) {
            mp.update();
        }
        Context3D ctx=this.scene3d.context3D;
//        material.fcData.printOut();
        ctx.setVc4fv(material.shader, "fc",material.fcNum, material.fcData.verBuff);
    }

    private void setCamPos(Material material) {

        material.updateCam(scene3d.camera3D.x / 100, scene3d.camera3D.y / 100, scene3d.camera3D.z / 100);
    }

    public TextureRes lightTextureRes;
    protected void setMaterialTexture(Material material, MaterialBaseParam mp)
    {

        Context3D ctx=this.scene3d.context3D;
        List<TexItem> texVec= mp.material.texList;
        TexItem texItem=null;
        for (int i   = 0; i < texVec.size(); i++) {
            texItem=texVec.get(i);
            if (texItem.isDynamic) {
                continue;
            }
            if (texItem.type == TexItem.LIGHTMAP) {
                if(lightTextureRes!=null){
                    ctx.setRenderTexture(material.shader,texItem.name,lightTextureRes.textTureInt,texItem.get_id());
                }
            }
            else if (texItem.type == TexItem.LTUMAP   ) {

                ctx.setRenderTexture(material.shader,texItem.name, Scene_data.pubLut.textTureInt,texItem.get_id());
            }
            else if (texItem.type == TexItem.CUBEMAP) {
               // Log.d(TAG, "CUBEMAP: ");

                if (material.useDynamicIBL) {// && _reflectionTextureVo) {
                    //_context.setTextureAt(texVec[i].id, _reflectionTextureVo.texture);
                } else {
                    float index   = (float)Math.floor(material.roughness * 5);
//                    if (Scene_data.skyCubeMap) {
//                        var cubeTexture: WebGLTexture = Scene_data.skyCubeMap[index];
//                        Scene_data.context3D.setRenderTextureCube($material.program, texVec[i].name, cubeTexture, texVec[i].id);
//                    }

                }
            }
            else if (texItem.type == 0) {
                ctx.setRenderTexture(material.shader,texItem.name,texItem.textureRes.textTureInt,texItem.get_id());
            }
        }
        List<DynamicBaseTexItem> texDynamicVec  =  mp.dynamicTexList;
        for (int i   = 0; i < texDynamicVec.size(); i++) {
            DynamicBaseTexItem dynamicBaseTexItem=texDynamicVec.get(i);
            texItem=(TexItem)dynamicBaseTexItem.target;
            if(texItem !=null&&dynamicBaseTexItem.textureRes!=null){
                ctx.setRenderTexture(material.shader,texItem.name,dynamicBaseTexItem.textureRes.textTureInt,texItem.get_id());
            }
        }

    }

    public void setBind(IBind $bindTarget, String $bindSocket) {
        this.bindTarget = $bindTarget;
        this.bindSocket = $bindSocket;
        this.bindMatrix = new Matrix3D();
    }

    public void setGroup(Vector3D $pos, Vector3D $rotaion, Vector3D $scale) {

        this._isInGroup = true;
        this._groupPos = $pos;
        this._groupRotation = $rotaion;
        this._groupScale = $scale;

        this.groupMatrix = new Matrix3D();
        this.groupRotationMatrix = new Matrix3D();

        this.groupMatrix.isIdentity = false;
        this.groupMatrix.identity();

        this.groupMatrix.appendScale($scale.x, $scale.y, $scale.z);
        this.groupMatrix.appendRotation($rotaion.x, Vector3D.X_AXIS);
        this.groupMatrix.appendRotation($rotaion.y, Vector3D.Y_AXIS);
        this.groupMatrix.appendRotation($rotaion.z, Vector3D.Z_AXIS);
        this.groupMatrix.appendTranslation($pos.x, $pos.y, $pos.z);

        this.groupRotationMatrix.isIdentity = false;
        this.groupRotationMatrix.identity();

        this.groupRotationMatrix.prependRotation($rotaion.z, Vector3D.Z_AXIS);
        this.groupRotationMatrix.prependRotation($rotaion.y, Vector3D.Y_AXIS);
        this.groupRotationMatrix.prependRotation($rotaion.x, Vector3D.X_AXIS);
    }
}
